tools/ds-identify

```diff
@@ -851,6 +851,11 @@ dscheck_NoCloud() {
         return ${DS_FOUND}
     fi
 
+    # This is a bit hacky, but a NoCloud false positive isn't the end of the world
+    if check_config "NoCloud" && check_config "user-data" && check_config "meta-data"; then
```

Comment:
`user-data` and `meta-data` are [required seed files](https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html) for NoCloud, but it looks like `check_config` searches `/etc/cloud/cloud.cfg` and `/etc/cloud/cloud.cfg.d/*` for keys with this name (am I reading that right?). Since I don't see those as keys in the doc, could you expand on why this works?

Reply:
Yeah, this functionality needs to be documented better. I just learned of this behavior due to the bug report :smile:  The only documentation is [an example](https://cloudinit.readthedocs.io/en/latest/topics/examples.html#configure-data-sources). The magic happens [in the datasource](https://github.com/canonical/cloud-init/blob/43c1cf433281844c30f0e4a19a938f4459bfccb7/cloudinit/sources/DataSourceNoCloud.py#L113). 

For the ds-identify piece, `check_config` just looks for a key at any nesting of the config (thanks bash). I'm basically just assuming that if 'NoCloud', 'user-data', and 'meta-data' exist somewhere in the config, then the datasource definition is happening there. If we're wrong, we waste a few cycles in python code checking for NoCloud.

Reply:
E.g., write create 
/etc/cloud/cloud.cfg.d/99_ds.cfg:
```yaml
datasource_list: ["NoCloud"]
datasource:
  NoCloud:
    user-data:  |
      #cloud-config
      runcmd: 
      - echo 'hi' > /var/tmp/hi 
    meta-data:
      instance-id: i-87018aed
      local-hostname: myhost.internal
```

Then `cloud-init clean --logs --reboot` and you should see those details reflected in your instance. With this branch, you shouldn't need the `datasource_list` line.

Reply:
Aha! Thanks. Yes this seems like something that should be documented better. I'm indifferent whether we should update the docs in this PR or a followup.

Reply:
Already in our docs tracking ticket